src/Tools/ExternalAccess/VisualDiagnostics/Internal/HotReloadDiagnosticSourceProvider.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics.Contracts;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics.Internal;
+
+internal abstract class HotReloadDiagnosticSourceProvider(IHotReloadDiagnosticManager diagnosticManager, bool isDocument) : IDiagnosticSourceProvider
+{
+    string IDiagnosticSourceProvider.Name => "HotReloadDiagnostics";
+    bool IDiagnosticSourceProvider.IsDocument => isDocument;
+
+    async ValueTask<ImmutableArray<IDiagnosticSource>> IDiagnosticSourceProvider.CreateDiagnosticSourcesAsync(RequestContext context, CancellationToken cancellationToken)
+    {
+        var hotReloadContext = new HotReloadRequestContext(context);
+        List<IDiagnosticSource> sources = new();
+        foreach (var provider in diagnosticManager.Providers)
+        {
+            if (provider.IsDocument == isDocument)
+            {
+                var hotReloadSources = await provider.CreateDiagnosticSourcesAsync(hotReloadContext, cancellationToken).ConfigureAwait(false);
+                sources.AddRange(hotReloadSources.Select(s => new HotReloadDiagnosticSource(s)));
+            }
+        }
+
+        var result = sources.ToImmutableArray();
+        return DiagnosticSourceManager.AggregateSourcesIfNeeded(result, isDocument);
```

Comment:
still don't understand aggregation.

Reply:
Comments added to AggregatedDocumentDiagnosticSource.